linera-core/src/client.rs

```diff
@@ -1294,6 +1294,189 @@ where
+        }
+
+        self.client.local_node.local_chain_info(chain_id).await
+    }
+
+    pub async fn try_synchronize_chain_state_from<A>(
+        &self,
+        name: ValidatorName,
+        mut node: A,
+        chain_id: ChainId,
+        notifications: &mut impl Extend<Notification>,
+    ) -> Result<(), ChainClientError>
+    where
+        A: LocalValidatorNode + Clone + 'static,
+    {
+        let local_info = self.client.local_node.local_chain_info(chain_id).await?;
+        let range = BlockHeightRange {
+            start: local_info.next_block_height,
+            limit: None,
+        };
+        let query = ChainInfoQuery::new(chain_id)
+            .with_sent_certificate_hashes_in_range(range)
+            .with_manager_values();
+        let info = match node.handle_chain_info_query(query).await {
+            Ok(response) if response.check(name).is_ok() => response.info,
+            Ok(_) => {
+                tracing::warn!("Ignoring invalid response from validator");
+                // Give up on this validator.
+                return Ok(());
+            }
+            Err(err) => {
+                tracing::warn!("Ignoring error from validator: {}", err);
+                return Ok(());
+            }
+        };
+
+        let certificates = future::try_join_all(
+            info.requested_sent_certificate_hashes
+                .into_iter()
+                .map(|hash| {
+                    let mut node = node.clone();
+                    async move { node.download_certificate(hash).await }
+                }),
+        )
+        .await?;
+
+        if !certificates.is_empty()
+            && self
+                .client
+                .local_node
+                .try_process_certificates(name, &mut node, chain_id, certificates, notifications)
+                .await
+                .is_none()
+        {
+            return Ok(());
+        };
+        if let Some(proposal) = info.manager.requested_proposed {
+            if proposal.content.block.chain_id == chain_id {
+                let owner = proposal.owner;
+                loop {
+                    let result = self
+                        .client
+                        .local_node
+                        .handle_block_proposal(*proposal.clone())
+                        .await;
+
+                    if let Err(original_err) = &result {
+                        if let LocalNodeError::WorkerError(WorkerError::ChainError(chain_error)) =
+                            original_err
+                        {
+                            if let ChainError::ExecutionError(
+                                ExecutionError::SystemError(
+                                    SystemExecutionError::BlobNotFoundOnRead(blob_id),
+                                ),
+                                _,
+                            ) = &**chain_error
+                            {
+                                self.update_lagging_validators(*blob_id).await?;
```

Comment:
Why do we need to update the lagging validators if our own local node is missing the blob? It would be clearer if `synchronize_from_validators` really only synchronized _from_, not _to_ them.

Reply:
I see your point, but AFAIU in our test we try to confirm a block that was validated by a different client, from a client without the blob, and also with lagging validators. How should we deal with this case if we're not updating the lagging validators and getting the blob while processing the certificate? I just couldn't find an alternative, but I'm probably missing something 🤔

Reply:
I'd expect the second client to only update the lagging validators when they send back an error about the missing blob.

Reply:
But this in the old version of the code won't send an error, right? If there's an error we don't return it, we just print a warning that we're skipping it and keep going here 🤔

Reply:
Had a chat about it: will just download the missing blobs if there are any